### PR TITLE
PRMP-1120 NRL Create pointer

### PR DESCRIPTION
**Pull request by NogaNHS**

### CODE REVIEW

File: lambdas/services/pds_api_service.py [hunk 1]
```diff
@@ -17,9 +15,9 @@
 logger = LoggingService(__name__)
 
 
-class PdsApiService(PatientSearch):
+class PdsApiService(PatientSearch, NhsOauthService):
```
Comment (steph-torres-nhs) [inline lambdas/services/pds_api_service.py:18]: Do we want to inherit, or compose/how do we want to handle auth?

File: lambdas/services/nrl_api_service.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        )
+
+    def create_new_pointer(self, body):
+        try:
+            self.set_x_request_id()
+            self.headers["Accept"] = "application/json"
+            response = self.session.post(
+                url=self.endpoint, headers=self.headers, json=body
+            )
+            response.raise_for_status()
```
Comment (steph-torres-nhs) [inline lambdas/services/nrl_api_service.py:47]: add log for success?

File: lambdas/services/nrl_api_service.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.endpoint = os.getenv("NRL_API_ENDPOINT")
+        self.session = requests.Session()
+        self.session.mount("https://", adapter)
+        self.end_user_ods_code = self._get_end_user_ods_code()
+        self.headers = {
+            "Authorization": f"Bearer {self.create_access_token()}",
+            "NHSD-End-User-Organisation-ODS": self.end_user_ods_code,
+        }
+
+    def _get_end_user_ods_code(self):
```
Comment (steph-torres-nhs) [inline lambdas/services/nrl_api_service.py:34]: "__" double underscore for private method

File: lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+
+def test_process_event_with_one_message(mock_service, context, set_env):
+    event = {"Records": [build_test_sqs_message("POST")]}
+
+    lambda_handler(event, context)
+
+    mock_service.create_new_pointer.assert_called_once()
+
```
Comment (steph-torres-nhs) [inline lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py:38]: test errors being thrown

File: lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            "url": "https://example.org/my-doc.pdf",
+        },
+    }
+    return {
+        "body": json.dumps(SQS_Message),
+        "eventSource": "aws:sqs",
+    }
+
+
+def test_process_event_with_one_message(mock_service, context, set_env):
```
Comment (steph-torres-nhs) [inline lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py:32]: is there a way to test success?

File: lambdas/services/base/nhs_oauth_service.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+from requests.exceptions import HTTPError
+from utils.audit_logging_setup import LoggingService
+from utils.exceptions import OAuthErrorException
+
+logger = LoggingService(__name__)
+
+
+class NhsOauthService:
+    def __init__(self, ssm_service):
+        self.ssm_service = ssm_service()
```
Comment (steph-torres-nhs) [inline lambdas/services/base/nhs_oauth_service.py:17]: callable?

File: lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+def test_process_delete_event_with_one_message(mock_service, context, set_env):
+    event = {"Records": [build_test_sqs_message("DELETE")]}
+
+    lambda_handler(event, context)
+
+    mock_service.delete_pointer.assert_called_once()
+
+
+def test_process_event_with_few_message(mock_service, context, set_env):
```
Comment (AndyFlintNHS) [inline lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py:56]: I'd rename `few_messages` to `multiple_messages`

Comment (AndyFlintNHS) [inline lambdas/tests/unit/handlers/test_manage_nrl_pointer_handler.py:56]: I'd rename `few_message` to `multiple_messages`

**Review by AndyFlintNHS — CHANGES_REQUESTED** [review]

**Review by steph-torres-nhs — APPROVED** [review]

**Pull request merged**